### PR TITLE
Typed Protocols: new API

### PR DESCRIPTION
**Pull request by coot**
# Description

Use `typed-protocols-0.3.0.0`.

Depends on:

* [x] input-output-hk/typed-protocols#52
* [x] input-output-hk/typed-protocols#61
* [x] IntersectMBO/ouroboros-network#4935

- **Updated to use typed-protocols-0.3.0.0**
- **Added KeepAlive tracer**

### CODE REVIEW

**Review by nfrisby — CHANGES_REQUESTED** [review]
The PR description seems a bit off. 
image

- It's version 0.3.0.0, not version 0.2.0.0.
- I don't see any aspects of `SizeInBytes` within the diff itself.

The diff itself looked totally fine. I don't follow the new `Stateful` stuff yet, but everything here seems very reasonable.

I'm not Approving because of:

1) The above concerns about the PR description.
2) I don't think we want an `source-repository-package` stanza on the `main` branch. Will the relevant dependencies be uploaded to CHaP soon enough to wait for them?

**Comment by coot** [comment]
I started with `t-p-0.2` and ended up with `t-p-0.3` :grin:.  The `TxSizeInBytes` patch is gone from this PR, due to consensus introducing its own sizes in a recent PR.

**Comment by crocodile-dentist** [comment]
ekg-forward 0.6 which is going into the next node release is incompatible with new typed-protocols 0.3 so lets hold off on merging this PR.

**Comment by coot** [comment]
@crocodile-dentist `ekg-forward` is not a dependency of `ouroboros-consensus`; I have a [draft PR](https://github.com/input-output-hk/ekg-forward/pull/33) for it as well (as for all other `cardano-node` dependencies :smiley:).

**Comment by crocodile-dentist** [comment]
> @crocodile-dentist `ekg-forward` is not a dependency of `ouroboros-consensus`; I have a [draft PR](https://github.com/input-output-hk/ekg-forward/pull/33) for it as well (as for all other `cardano-node` dependencies 😃).

Yes, but it would have caused problems downstream when releasing node - consensus would have to release off branch to make it go smoothest.

**Review by amesgen — APPROVED** [review]

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/IOLike.hs [hunk 1]
```diff
@@ -68,6 +68,7 @@ import           Ouroboros.Consensus.Util.Orphans ()
 -------------------------------------------------------------------------------}
 
 class ( MonadAsync              m
+      , MonadLabelledSTM        m
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/IOLike.hs:71]: It seems impossible to define `instance MonadTraceSTM (OverrideDelay m)` right now.  One needs:
```fn
fn :: proxy (OverrideDelay m) -> proxy m
fn = undefined
```

We need to switch from a polymorphic `proxy` type to `Proxy` in `MonadTraceTVar`.

Here are the definitions which compiled for me:
```hs
instance MonadInspectSTM m => MonadInspectSTM (OverrideDelay m) where
  type InspectMonad (OverrideDelay m) = InspectMonad m
  inspectTVar p v = inspectTVar p v
  inspectTMVar p v = inspectTMVar p v

fn :: proxy (OverrideDelay m) -> proxy m
fn = undefined

instance MonadTraceSTM m => MonadTraceSTM (OverrideDelay m) where
  traceTVar p v f  = OverrideDelaySTM . lift $ LazySTM.traceTVar (fn p) v f
  traceTMVar p v f = OverrideDelaySTM . lift $ LazySTM.traceTMVar (fn p) v f
  traceTQueue p v f = OverrideDelaySTM . lift $ LazySTM.traceTQueue (fn p) v f
  traceTBQueue p v f = OverrideDelaySTM . lift $ LazySTM.traceTBQueue (fn p) v f
  traceTSem p v f = OverrideDelaySTM . lift $ LazySTM.traceTSem (fn p) v f
```

If we used `Proxy`, then `fn Proxy = Proxy` should work.

[the remark this answers is not in the document]
Comment (amesgen) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/IOLike.hs:71]: But the `proxy`s don't need to be the same, right?
```haskell
fn :: proxy (OverrideDelay m) -> Proxy m
fn _ = Proxy
```
should do the job.

[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/IOLike.hs:71]: yes, I realised that when I was driving... :grin:

File: ouroboros-consensus/test/consensus-test/Test/Consensus/BlockchainTime/Simple.hs [hunk 2]
```diff
@@ -539,6 +539,18 @@ instance MonadLabelledSTM m => MonadLabelledSTM (OverrideDelay m) where
   labelTQueueIO v  = OverrideDelay . lift . LazySTM.labelTQueueIO v
   labelTBQueueIO v = OverrideDelay . lift . LazySTM.labelTBQueueIO v
 
+instance MonadInspectSTM m => MonadInspectSTM (OverrideDelay m) where
+  type InspectMonad (OverrideDelay m) = InspectMonad m
+  inspectTVar = inspectTVar
+  inspectTMVar = inspectTMVar
```
Comment (amesgen) [inline ouroboros-consensus/test/consensus-test/Test/Consensus/BlockchainTime/Simple.hs:545]: Won't this loop at runtime? I think we need to supply `Proxy @m`

**Pull request merged**